src/main/java/org/traccar/protocol/ZrProtocolDecoder.java

```diff
@@ -0,0 +1,323 @@
+                        deviceSession.set(STATE_ACC, ACC_OFF);
+                    }
+                    break;
+            }
+        }
+
+        if (position.getLatitude() != 0 && position.getLongitude() != 0) {
+            position.setValid(true);
+        } else {
+            return null;
+        }
+
+        Integer accState = deviceSession.get(STATE_ACC);
+        position.set(Position.KEY_IGNITION, accState != null && accState.equals(ACC_ON));
+        return position;
+    }
+
+    private String decode2310(long value) {
+        if (BitUtil.check(value, 3)) {
+            return Position.ALARM_POWER_OFF;
+        }
+        if (BitUtil.check(value, 4)) {
+            return Position.ALARM_LOW_POWER;
+        }
+        if (BitUtil.check(value, 22)) {
+            return Position.ALARM_OVERSPEED;
+        }
+        if (BitUtil.check(value, 41)) {
+            return Position.ALARM_TOW;
+        }
+        if (BitUtil.check(value, 42)) {
+            return Position.ALARM_FALL_DOWN;
+        }
+        if (BitUtil.check(value, 43)) {
+            return Position.ALARM_ACCIDENT;
+        }
+        if (BitUtil.check(value, 44)) {
+            return Position.ALARM_VIBRATION;
+        }
+        if (BitUtil.check(value, 50)) {
+            return Position.ALARM_REMOVING;
+        }
+
+        return null;
+    }
+
+    private String decodeId(ByteBuf id) {
+        String str = ByteBufUtil.hexDump(id);
+        int i = 0;
+        while (i < str.length() && str.charAt(i) == '0') {
+            i++;
+        }
+        String devId = str.substring(i);
+        // if the devId is all zero return default 0
+        if (devId.length() == 0) {
+            return "0";
+        }
+        return devId;
+    }
+
+    public static ByteBuf formatMessage(int type, ByteBuf id, short editionNum, byte encryptionType, Integer packageNo, ByteBuf body) {
+        ByteBuf buffer = Unpooled.buffer();
+
+        // header
+        buffer.writeBytes(Unpooled.wrappedBuffer(ByteBufUtil.decodeHexDump(FRAME_HEADER)));
+        // Message length, please specify one at will
+        buffer.writeShort(12);
+        // Protocol version number
+        buffer.writeShort(editionNum);
+        // Encryption type
+        buffer.writeByte(encryptionType);
+        // Terminal device number
+        buffer.writeBytes(id);
+        // msg id
+        buffer.writeShort(type);
+        // subcontract item
+        buffer.writeByte(0x11);
+        // packageNo
+        buffer.writeShort(packageNo);
+        // body len
+        buffer.writeByte(body.readableBytes());
+        // body
+        buffer.writeBytes(body);
+        // check sum
+        buffer.writeByte(12);
+        // msg tail
+        buffer.writeBytes(Unpooled.wrappedBuffer(ByteBufUtil.decodeHexDump(FRAME_TAIL)));
+
+        // Correct message length
+        buffer.setShort(2, buffer.readableBytes() - 6);
+        buffer.setByte(buffer.writerIndex() - 3, bcc(buffer));
```

Comment:
Why can't we calculate it from the start?

Reply:
This is a protocol requirement.

Reply:
Because at the beginning, in the message header, we are still unable to calculate the exact message length.